File: server/controllers/ragController.ts
import { Request, Response } from 'express';
import { ragService } from '../services/ragService.ts';
import { logger } from '../services/logger.ts';

export const getConversations = async (req: Request, res: Response) => {
  try {
    const list = ragService.getAllConversations();
    return res.json({
      total: list.length,
      conversations: list,
    });
  } catch (err) {
    logger.error('Failed to get RAG conversations', { error: String(err) });
    return res.status(500).json({ error: 'Failed to retrieve knowledge base conversations' });
  }
};

export const addConversation = async (req: Request, res: Response) => {
  try {
    const { title, category, tags, level, rawTranscript, turns, locationOrContext, worldContextTopic, source } = req.body;

    if (!rawTranscript || !rawTranscript.trim()) {
      return res.status(400).json({ error: 'rawTranscript is required' });
    }

    const record = await ragService.addConversationRecord({
      title,
      category,
      tags,
      level,
      rawTranscript,
      turns,
      locationOrContext,
      worldContextTopic,
      source: source || 'live_recorded',
    });

    return res.status(201).json({
      message: 'Gesprek succesvol verwerkt en toegevoegd aan de RAG vector kennisbank!',
      record,
    });
  } catch (err) {
    logger.error('Failed to ingest conversation into RAG knowledge base', { error: String(err) });
    return res.status(500).json({ error: 'Failed to process and index conversation' });
  }
};

export const deleteConversation = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const success = ragService.deleteConversation(id);
    if (!success) {
      return res.status(404).json({ error: 'Conversation record not found' });
    }
    return res.json({ message: 'Gesprek verwijderd uit de kennisbank' });
  } catch (err) {
    return res.status(500).json({ error: 'Failed to delete conversation' });
  }
};

export const searchRAG = async (req: Request, res: Response) => {
  try {
    const { query, category, topK = 3 } = req.body;
    if (!query || !query.trim()) {
      return res.status(400).json({ error: 'Query is required' });
    }

    const matches = await ragService.searchKnowledgeBase(query, Number(topK) || 3, category);
    return res.json({
      query,
      matchCount: matches.length,
      matches,
    });
  } catch (err) {
    logger.error('RAG vector search failed', { error: String(err) });
    return res.status(500).json({ error: 'Failed to perform semantic search' });
  }
};

export const getWorldTopics = async (req: Request, res: Response) => {
  try {
    const topics = ragService.getAllWorldTopics();
    return res.json({ topics });
  } catch (err) {
    return res.status(500).json({ error: 'Failed to get world topics' });
  }
};

export const toggleWorldTopic = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const active = ragService.toggleWorldTopic(id);
    return res.json({ topicId: id, activeInRAG: active });
  } catch (err) {
    return res.status(500).json({ error: 'Failed to toggle world topic' });
  }
};

export const getVectorDbStats = async (req: Request, res: Response) => {
  try {
    const stats = await ragService.getVectorDbStats();
    return res.json(stats);
  } catch (err) {
    logger.error('Failed to get vector database stats', { error: String(err) });
    return res.status(500).json({ error: 'Failed to fetch vector DB statistics' });
  }
};

File: backend/src/main/java/com/praatnederlands/controller/RagController.java
package com.praatnederlands.controller;

import com.praatnederlands.model.GlobalWorldTopic;
import com.praatnederlands.model.RAGSearchMatch;
import com.praatnederlands.model.RealtimeConversationRecord;
import com.praatnederlands.service.RagService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/rag")
public class RagController {

    private final RagService ragService;

    public record SearchRequest(String query, Integer topK, String category) {}

    public RagController(RagService ragService) {
        this.ragService = ragService;
    }

    @GetMapping("/conversations")
    public ResponseEntity<Map<String, Object>> getConversations() {
        List<RealtimeConversationRecord> list = ragService.getAllConversations();
        return ResponseEntity.ok(Map.of(
            "conversations", list,
            "count", list.size(),
            "timestamp", java.time.Instant.now().toString()
        ));
    }

    @PostMapping("/conversations")
    public ResponseEntity<Map<String, Object>> addConversation(@RequestBody RealtimeConversationRecord record) {
        if (record == null || record.rawTranscript() == null || record.rawTranscript().isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Transcript is required"));
        }
        RealtimeConversationRecord saved = ragService.addConversationRecord(record);
        return ResponseEntity.ok(Map.of(
            "success", true,
            "conversation", saved
        ));
    }

    @DeleteMapping("/conversations/{id}")
    public ResponseEntity<Map<String, Object>> deleteConversation(@PathVariable String id) {
        boolean removed = ragService.deleteConversation(id);
        return ResponseEntity.ok(Map.of("success", removed));
    }

    @PostMapping("/search")
    public ResponseEntity<Map<String, Object>> searchRAG(@RequestBody SearchRequest req) {
        String query = req != null ? req.query() : "";
        int topK = (req != null && req.topK() != null) ? req.topK() : 5;
        String category = req != null ? req.category() : "all";

        List<RAGSearchMatch> matches = ragService.searchKnowledgeBase(query, topK, category);
        return ResponseEntity.ok(Map.of(
            "query", query,
            "matches", matches,
            "count", matches.size()
        ));
    }

    @GetMapping("/world-topics")
    public ResponseEntity<Map<String, Object>> getWorldTopics() {
        List<GlobalWorldTopic> topics = ragService.getAllWorldTopics();
        return ResponseEntity.ok(Map.of("topics", topics));
    }

    @PostMapping("/world-topics/{id}/toggle")
    public ResponseEntity<Map<String, Object>> toggleTopic(@PathVariable String id) {
        boolean toggled = ragService.toggleWorldTopic(id);
        return ResponseEntity.ok(Map.of("success", toggled));
    }
}
